File: research/verdict_research/features/feature_vector.py
import re
from dataclasses import dataclass
from datetime import UTC, date, datetime

from verdict_research.features.listing_drift import (
    ListingDriftResult,
    ReviewForDrift,
    listing_identity_drift,
)
from verdict_research.features.rating_deconvolution import (
    RatingDeconvolutionResult,
    rating_deconvolution,
)
from verdict_research.features.reviewer_graph import (
    ReviewerGraphResult,
    ReviewForReviewerGraph,
    reviewer_graph_share,
)
from verdict_research.features.temporal_burst import (
    Burst,
    TemporalBurstResult,
    detect_temporal_bursts,
)
from verdict_research.features.text_near_duplication import (
    ReviewForNearDuplication,
    TextNearDuplicationResult,
    text_near_duplication,
)
from verdict_research.features.verification_concentration import (
    ReviewForVerification,
    VerificationConcentrationResult,
    verification_concentration,
)
from verdict_research.schema import Review

MINIMUM_REVIEW_COUNT = 30
MINIMUM_DATED_REVIEW_COUNT = 20
MINIMUM_HISTORY_DAYS = 21

_EPOCH_ORDINAL = date(1970, 1, 1).toordinal()
_SECONDS_PER_DAY = 86_400
_ISO_DATE_ONLY = re.compile(r"\d{4}-\d{2}-\d{2}")
_HAS_EXPLICIT_ZONE = re.compile(r"Z$|[+-]\d{2}:\d{2}$")


def day_index(iso: str) -> int:
    if "T" not in iso:
        if not _ISO_DATE_ONLY.fullmatch(iso):
            raise ValueError(f"day_index requires an iso date with no ambiguous zone: {iso}")
        year, month, day = int(iso[0:4]), int(iso[5:7]), int(iso[8:10])
        return date(year, month, day).toordinal() - _EPOCH_ORDINAL
    if not _HAS_EXPLICIT_ZONE.search(iso):
        raise ValueError(f"day_index requires an explicit time zone on a datetime string: {iso}")
    dt = datetime.fromisoformat(iso.replace("Z", "+00:00"))
    epoch_seconds = dt.astimezone(UTC).timestamp()
    return int(epoch_seconds // _SECONDS_PER_DAY)


def meets_minimum_data_thresholds(reviews: list[Review]) -> bool:
    if len(reviews) < MINIMUM_REVIEW_COUNT:
        return False
    dated_days = [day_index(r.date) for r in reviews if r.date is not None]
    if len(dated_days) < MINIMUM_DATED_REVIEW_COUNT:
        return False
    return max(dated_days) - min(dated_days) >= MINIMUM_HISTORY_DAYS


def build_rating_histogram(reviews: list[Review]) -> list[float] | None:
    rated = [r for r in reviews if r.rating is not None]
    if not rated:
        return None
    bins = [0, 0, 0, 0, 0]
    for review in rated:
        bin_index = min(5, max(1, round(review.rating))) - 1
        bins[bin_index] += 1
    return [count / len(rated) for count in bins]


def derive_day_indices(reviews: list[Review]) -> list[int | None]:
    return [None if r.date is None else day_index(r.date) for r in reviews]


@dataclass
class DailyCounts:
    daily_counts: list[int]
    min_day: int


def build_daily_counts(reviews: list[Review]) -> DailyCounts | None:
    days = [day_index(r.date) for r in reviews if r.date is not None]
    if not days:
        return None
    min_day, max_day = min(days), max(days)
    daily_counts = [0] * (max_day - min_day + 1)
    for day in days:
        daily_counts[day - min_day] += 1
    return DailyCounts(daily_counts=daily_counts, min_day=min_day)


def derive_inside_burst(reviews: list[Review], min_day: int, bursts: list[Burst]) -> list[bool]:
    result = []
    for review in reviews:
        if review.date is None:
            result.append(False)
            continue
        day = day_index(review.date) - min_day
        result.append(any(b.start_day <= day <= b.end_day for b in bursts))
    return result


@dataclass
class FeatureVectorInputs:
    organic_prior: list[float]
    injection_kernel: list[float]
    window_days: int = 28
    percentile: float = 0.99
    product_text: str = ""
    flagged_reviewer_ids: set[str] | None = None


@dataclass
class FeatureVector:
    meets_minimum_data: bool
    rating_deconvolution: RatingDeconvolutionResult | None
    temporal_burst: TemporalBurstResult | None
    verification_concentration: VerificationConcentrationResult | None
    text_near_duplication: TextNearDuplicationResult
    listing_drift: ListingDriftResult
    reviewer_graph: ReviewerGraphResult | None


def build_feature_vector(reviews: list[Review], inputs: FeatureVectorInputs) -> FeatureVector:
    meets_minimum_data = meets_minimum_data_thresholds(reviews)

    histogram = build_rating_histogram(reviews)
    rating_result = (
        rating_deconvolution(histogram, inputs.organic_prior, inputs.injection_kernel)
        if histogram is not None
        else None
    )

    daily = build_daily_counts(reviews)
    temporal_result = (
        detect_temporal_bursts(daily.daily_counts, inputs.window_days, inputs.percentile)
        if daily is not None
        else None
    )

    verification_result = None
    if daily is not None and temporal_result is not None:
        inside_burst = derive_inside_burst(reviews, daily.min_day, temporal_result.bursts)
        verification_result = verification_concentration(
            [
                ReviewForVerification(
                    rating=review.rating, verified=review.verified, inside_burst=inside_burst[i]
                )
                for i, review in enumerate(reviews)
            ]
        )

    duplication_result = text_near_duplication(
        [ReviewForNearDuplication(text=review.text) for review in reviews]
    )

    drift_result = listing_identity_drift(
        [ReviewForDrift(text=review.text) for review in reviews],
        derive_day_indices(reviews),
        inputs.product_text,
    )

    reviewer_graph_result = (
        None
        if inputs.flagged_reviewer_ids is None
        else reviewer_graph_share(
            [ReviewForReviewerGraph(reviewer_id=review.reviewer_id) for review in reviews],
            inputs.flagged_reviewer_ids,
        )
    )

    return FeatureVector(
        meets_minimum_data=meets_minimum_data,
        rating_deconvolution=rating_result,
        temporal_burst=temporal_result,
        verification_concentration=verification_result,
        text_near_duplication=duplication_result,
        listing_drift=drift_result,
        reviewer_graph=reviewer_graph_result,
    )

File: extension/src/score/featureVector.ts
import type { Review } from "../extract/types";
import {
  listingIdentityDrift,
  type ListingDriftOptions,
  type ListingDriftResult,
} from "./listingDrift";
import { ratingDeconvolution, type RatingDeconvolutionResult } from "./ratingDeconvolution";
import { reviewerGraphShare, type ReviewerGraphResult } from "./reviewerGraph";
import { detectTemporalBursts, type Burst, type TemporalBurstResult } from "./temporalBurst";
import {
  textNearDuplication,
  type TextNearDuplicationOptions,
  type TextNearDuplicationResult,
} from "./textNearDuplication";
import {
  verificationConcentration,
  type VerificationConcentrationResult,
} from "./verificationConcentration";

// SPEC.md section 6 minimum data thresholds
export const MINIMUM_REVIEW_COUNT = 30;
export const MINIMUM_DATED_REVIEW_COUNT = 20;
export const MINIMUM_HISTORY_DAYS = 21;

const MS_PER_DAY = 86_400_000;

// a zoneless date parses per reader
const ISO_DATE_ONLY = /^\d{4}-\d{2}-\d{2}$/;
const ISO_DATETIME_WITH_ZONE = /^\d{4}-\d{2}-\d{2}T[\d:.]+(?:Z|[+-]\d{2}:\d{2})$/;

export function dayIndex(iso: string): number {
  if (!ISO_DATE_ONLY.test(iso) && !ISO_DATETIME_WITH_ZONE.test(iso)) {
    throw new Error(`dayIndex requires an iso date with no ambiguous zone: ${iso}`);
  }
  return Math.floor(Date.parse(iso) / MS_PER_DAY);
}

export function meetsMinimumDataThresholds(reviews: readonly Review[]): boolean {
  if (reviews.length < MINIMUM_REVIEW_COUNT) {
    return false;
  }
  const datedDays = reviews
    .filter((review): review is Review & { date: string } => review.date !== null)
    .map((review) => dayIndex(review.date));
  if (datedDays.length < MINIMUM_DATED_REVIEW_COUNT) {
    return false;
  }
  const span = Math.max(...datedDays) - Math.min(...datedDays);
  return span >= MINIMUM_HISTORY_DAYS;
}

export function buildRatingHistogram(reviews: readonly Review[]): number[] | null {
  const rated = reviews.filter((review): review is Review & { rating: number } =>
    review.rating !== null
  );
  if (rated.length === 0) {
    return null;
  }
  const bins = [0, 0, 0, 0, 0];
  for (const review of rated) {
    const bin = Math.min(5, Math.max(1, Math.round(review.rating))) - 1;
    bins[bin] = (bins[bin] ?? 0) + 1;
  }
  return bins.map((count) => count / rated.length);
}

export function deriveDayIndices(reviews: readonly Review[]): (number | null)[] {
  return reviews.map((review) => (review.date === null ? null : dayIndex(review.date)));
}

export interface DailyCounts {
  dailyCounts: number[];
  minDay: number;
}

export function buildDailyCounts(reviews: readonly Review[]): DailyCounts | null {
  const days = reviews
    .filter((review): review is Review & { date: string } => review.date !== null)
    .map((review) => dayIndex(review.date));
  if (days.length === 0) {
    return null;
  }
  const minDay = Math.min(...days);
  const maxDay = Math.max(...days);
  const dailyCounts = new Array<number>(maxDay - minDay + 1).fill(0);
  for (const day of days) {
    const index = day - minDay;
    dailyCounts[index] = (dailyCounts[index] ?? 0) + 1;
  }
  return { dailyCounts, minDay };
}

export function deriveInsideBurst(
  reviews: readonly Review[],
  minDay: number,
  bursts: readonly Burst[],
): boolean[] {
  return reviews.map((review) => {
    if (review.date === null) {
      return false;
    }
    const day = dayIndex(review.date) - minDay;
    return bursts.some((burst) => day >= burst.startDay && day <= burst.endDay);
  });
}

export interface FeatureVectorInputs {
  organicPrior: readonly number[];
  injectionKernel: readonly number[];
  windowDays?: number;
  percentile?: number;
  textNearDuplicationSignatureCache?: TextNearDuplicationOptions["signatureCache"];
  textNearDuplicationLinkCache?: TextNearDuplicationOptions["linkCache"];
  productText?: string;
  listingDriftEmbeddingCache?: ListingDriftOptions["embeddingCache"];
  flaggedReviewerIds?: ReadonlySet<string>;
}

export interface FeatureVector {
  meetsMinimumData: boolean;
  ratingDeconvolution: RatingDeconvolutionResult | null;
  temporalBurst: TemporalBurstResult | null;
  verificationConcentration: VerificationConcentrationResult | null;
  textNearDuplication: TextNearDuplicationResult;
  listingDrift: ListingDriftResult;
  reviewerGraph: ReviewerGraphResult | null;
}

export function buildFeatureVector(
  reviews: readonly Review[],
  inputs: FeatureVectorInputs,
): FeatureVector {
  const meetsMinimumData = meetsMinimumDataThresholds(reviews);

  const histogram = buildRatingHistogram(reviews);
  const ratingResult = histogram !== null
    ? ratingDeconvolution(histogram, inputs.organicPrior, inputs.injectionKernel)
    : null;

  const daily = buildDailyCounts(reviews);
  const temporalResult = daily !== null
    ? detectTemporalBursts(daily.dailyCounts, inputs.windowDays, inputs.percentile)
    : null;

  let verificationResult: VerificationConcentrationResult | null = null;
  if (daily !== null && temporalResult !== null) {
    const insideBurst = deriveInsideBurst(reviews, daily.minDay, temporalResult.bursts);
    verificationResult = verificationConcentration(
      reviews.map((review, i) => ({
        rating: review.rating,
        verified: review.verified,
        insideBurst: insideBurst[i] ?? false,
      })),
    );
  }

  const duplicationResult = textNearDuplication(reviews, {
    signatureCache: inputs.textNearDuplicationSignatureCache,
    linkCache: inputs.textNearDuplicationLinkCache,
  });

  const driftResult = listingIdentityDrift(
    reviews,
    deriveDayIndices(reviews),
    inputs.productText ?? "",
    { embeddingCache: inputs.listingDriftEmbeddingCache },
  );

  const reviewerGraphResult = inputs.flaggedReviewerIds !== undefined
    ? reviewerGraphShare(reviews, inputs.flaggedReviewerIds)
    : null;

  return {
    meetsMinimumData,
    ratingDeconvolution: ratingResult,
    temporalBurst: temporalResult,
    verificationConcentration: verificationResult,
    textNearDuplication: duplicationResult,
    listingDrift: driftResult,
    reviewerGraph: reviewerGraphResult,
  };
}

function asRecord(value: unknown): Record<string, unknown> | null {
  return typeof value === "object" && value !== null && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : null;
}

function finite(value: unknown): number | null {
  return typeof value === "number" && Number.isFinite(value) ? value : null;
}

function nullableFinite(value: unknown): number | null | undefined {
  if (value === null) {
    return null;
  }
  return finite(value) ?? undefined;
}

function allPresent(values: readonly (number | null | undefined)[]): boolean {
  return values.every((value) => value !== undefined && value !== null);
}

// a vector read out of a file was written by nobody we can vouch for, and an absent
// feature imputed as zero is a number nobody measured, so a bad one is refused whole
export function parseFeatureVector(value: unknown): FeatureVector | null {
  const record = asRecord(value);
  if (record === null || typeof record.meetsMinimumData !== "boolean") {
    return null;
  }

  const duplication = asRecord(record.textNearDuplication);
  const drift = asRecord(record.listingDrift);
  if (duplication === null || drift === null) {
    return null;
  }

  const clusterCount = finite(duplication.clusterCount);
  const largestClusterShare = finite(duplication.largestClusterShare);
  const offTopicCount = finite(drift.offTopicCount);
  const driftStatistic = finite(drift.driftStatistic);
  const embeddedCount = finite(drift.embeddedCount);
  if (!allPresent([clusterCount, largestClusterShare, offTopicCount, driftStatistic, embeddedCount])) {
    return null;
  }

  const duplicateReviewShare = nullableFinite(duplication.duplicateReviewShare);
  const offTopicShare = nullableFinite(drift.offTopicShare);
  const meanDistance = nullableFinite(drift.meanDistance);
  if (duplicateReviewShare === undefined || offTopicShare === undefined || meanDistance === undefined) {
    return null;
  }

  const rating = parseRatingDeconvolution(record.ratingDeconvolution);
  const burst = parseTemporalBurst(record.temporalBurst);
  const verification = parseVerificationConcentration(record.verificationConcentration);
  const graph = parseReviewerGraph(record.reviewerGraph);
  if (rating === undefined || burst === undefined || verification === undefined || graph === undefined) {
    return null;
  }

  return {
    meetsMinimumData: record.meetsMinimumData,
    ratingDeconvolution: rating,
    temporalBurst: burst,
    verificationConcentration: verification,
    textNearDuplication: {
      duplicateReviewShare,
      clusterCount: clusterCount as number,
      largestClusterShare: largestClusterShare as number,
    },
    listingDrift: {
      offTopicShare,
      offTopicCount: offTopicCount as number,
      meanDistance,
      // nothing reads these back, so they do not cross a file boundary
      changePoint: null,
      driftStatistic: driftStatistic as number,
      embeddedCount: embeddedCount as number,
    },
    reviewerGraph: graph,
  };
}

// undefined means the field was there and unreadable, which fails the whole vector
function parseRatingDeconvolution(value: unknown): RatingDeconvolutionResult | null | undefined {
  if (value === null || value === undefined) {
    return null;
  }
  const record = asRecord(value);
  const injectedShare = finite(record?.injectedShare);
  const residualError = finite(record?.residualError);
  return allPresent([injectedShare, residualError])
    ? { injectedShare: injectedShare as number, residualError: residualError as number }
    : undefined;
}

function parseTemporalBurst(value: unknown): TemporalBurstResult | null | undefined {
  if (value === null || value === undefined) {
    return null;
  }
  const record = asRecord(value);
  const burstFraction = finite(record?.burstFraction);
  const burstCount = finite(record?.burstCount);
  const largestBurstShare = finite(record?.largestBurstShare);
  return allPresent([burstFraction, burstCount, largestBurstShare])
    ? {
      bursts: [],
      burstFraction: burstFraction as number,
      burstCount: burstCount as number,
      largestBurstShare: largestBurstShare as number,
    }
    : undefined;
}

function parseVerificationConcentration(
  value: unknown,
): VerificationConcentrationResult | null | undefined {
  if (value === null || value === undefined) {
    return null;
  }
  const record = asRecord(value);
  const lift = nullableFinite(record?.lift);
  const baseCount = finite(record?.baseCount);
  if (lift === undefined || baseCount === null) {
    return undefined;
  }
  return { lift, baseCount };
}

function parseReviewerGraph(value: unknown): ReviewerGraphResult | null | undefined {
  if (value === null || value === undefined) {
    return null;
  }
  const record = asRecord(value);
  const flaggedReviewShare = nullableFinite(record?.flaggedReviewShare);
  const flaggedReviewCount = finite(record?.flaggedReviewCount);
  const flaggedReviewerCount = finite(record?.flaggedReviewerCount);
  const knownReviewerCount = finite(record?.knownReviewerCount);
  const identifiedReviewCount = finite(record?.identifiedReviewCount);
  if (
    flaggedReviewShare === undefined ||
    !allPresent([flaggedReviewCount, flaggedReviewerCount, knownReviewerCount, identifiedReviewCount])
  ) {
    return undefined;
  }
  return {
    flaggedReviewShare,
    flaggedReviewCount: flaggedReviewCount as number,
    flaggedReviewerCount: flaggedReviewerCount as number,
    knownReviewerCount: knownReviewerCount as number,
    identifiedReviewCount: identifiedReviewCount as number,
  };
}
